File: javascript/javascript/Python3/JavaScriptParserBase.py

from antlr4 import *

relativeImport = False
if __name__ is not None and "." in __name__:
    relativeImport = True

class JavaScriptParserBase(Parser):
    @staticmethod
    def parser():
        if relativeImport:
            from .JavaScriptParser import JavaScriptParser
        else:
            from JavaScriptParser import JavaScriptParser
        return JavaScriptParser

    def p(self, s : str) -> bool:
        return self.prev(s)

    def prev(self, s : str) -> bool:
        return self._input.LT(-1).text == s

    def n(self, s : str) -> bool:
        return self.next(s)

    def next(self, s: str) -> bool:
        return self._input.LT(1).text == s

    def notLineTerminator(self) -> bool:
        JavaScriptParser = self.parser()

        return not self.here(JavaScriptParser.LineTerminator)

    def notOpenBraceAndNotFunction(self) -> bool:
        JavaScriptParser = self.parser()

        nextTokenType = self._input.LT(1).type
        return nextTokenType != JavaScriptParser.OpenBrace and nextTokenType != JavaScriptParser.Function_

    def closeBrace(self) -> bool:
        JavaScriptParser = self.parser()

        return self._input.LT(1).type == JavaScriptParser.CloseBrace

    def here(self, tokenType: int) -> bool:
        """
        Returns {@code true} iff on the current index of the parser's
        token stream a token of the given {@code type} exists on the
        {@code HIDDEN} channel.
        :param:type:
                   the type of the token on the {@code HIDDEN} channel
                   to check.
        :return:{@code true} iff on the current index of the parser's
            token stream a token of the given {@code type} exists on the
            {@code HIDDEN} channel.
        """
        # Get the token ahead of the current index.
        assert isinstance(self.getCurrentToken(), Token)
        possibleIndexEosToken: Token = self.getCurrentToken().tokenIndex - 1
        ahead = self._input.get(possibleIndexEosToken)

        # Check if the token resides on the HIDDEN channel and if it's of the
        # provided type.
        return (ahead.channel == Lexer.HIDDEN) and (ahead.type == tokenType)

    def lineTerminatorAhead(self) -> bool:
        """
        Returns {@code true} iff on the current index of the parser's
        token stream a token exists on the {@code HIDDEN} channel which
        either is a line terminator, or is a multi line comment that
        contains a line terminator.

        :return: {@code true} iff on the current index of the parser's
        token stream a token exists on the {@code HIDDEN} channel which
        either is a line terminator, or is a multi line comment that
        contains a line terminator.
        """
        JavaScriptParser = self.parser()

        # Get the token ahead of the current index.
        possibleIndexEosToken: Token = self.getCurrentToken().tokenIndex - 1
        ahead: Token = self._input.get(possibleIndexEosToken)

        if ahead.channel != Lexer.HIDDEN:
            # We're only interested in tokens on the HIDDEN channel.
            return False

        if ahead.type == JavaScriptParser.LineTerminator:
            # There is definitely a line terminator ahead.
            return True

        if ahead.type == JavaScriptParser.WhiteSpaces:
            # Get the token ahead of the current whitespaces.
            possibleIndexEosToken = self.getCurrentToken().tokenIndex - 2
            ahead = self._input.get(possibleIndexEosToken)

        # Get the token's text and type.
        text = ahead.text
        tokenType = ahead.type

        # Check if the token is, or contains a line terminator.
        return ((tokenType == JavaScriptParser.MultiLineComment and (text.contains("\r") or text.contains("\n"))) or
                (tokenType == JavaScriptParser.LineTerminator))

File: javascript/javascript/Go/javascript_parser_base.go
package parser

import (
	"strings"

	"github.com/antlr4-go/antlr/v4"
)

// JavaScriptParserBase implementation.
type JavaScriptParserBase struct {
	*antlr.BaseParser
}

// Short for p.prev(str string)
func (p *JavaScriptParserBase) p(str string) bool {
	return p.prev(str)
}

// Whether the previous token value equals to str.
func (p *JavaScriptParserBase) prev(str string) bool {
	return p.GetTokenStream().LT(-1).GetText() == str
}

// Short for p.next(str string)
func (p *JavaScriptParserBase) n(str string) bool {
	return p.next(str)
}

// Whether the next token value equals to str.
func (p *JavaScriptParserBase) next(str string) bool {
	return p.GetTokenStream().LT(1).GetText() == str
}

func (p *JavaScriptParserBase) notLineTerminator() bool {
	return !p.here(JavaScriptParserLineTerminator)
}

func (p *JavaScriptParserBase) notOpenBraceAndNotFunction() bool {
	nextTokenType := p.GetTokenStream().LT(1).GetTokenType()
	return nextTokenType != JavaScriptParserOpenBrace && nextTokenType != JavaScriptParserFunction_
}

func (p *JavaScriptParserBase) closeBrace() bool {
	return p.GetTokenStream().LT(1).GetTokenType() == JavaScriptParserCloseBrace
}

// Returns true if on the current index of the parser's
// token stream a token of the given type exists on the
// Hidden channel.
func (p *JavaScriptParserBase) here(_type int) bool {
	// Get the most recently emitted token.
	currentToken := p.GetTokenStream().LT(-1)

	// Get the next token index.
	nextTokenIndex := 0

	if currentToken != nil {
		nextTokenIndex = currentToken.GetTokenIndex() + 1
	}

	// Get the token after the `currentToken`. By using `_input.get(index)`,
	// we also grab a token that is (possibly) on the HIDDEN channel.
	nextToken := p.GetTokenStream().Get(nextTokenIndex);

	// Check if the token resides on the HIDDEN channel and if it's of the
	// provided type.
	return nextToken.GetChannel() == antlr.LexerHidden && nextToken.GetTokenType() == _type
}

// Returns true if on the current index of the parser's
// token stream a token exists on the Hidden channel which
// either is a line terminator, or is a multi line comment that
// contains a line terminator.
func (p *JavaScriptParserBase) lineTerminatorAhead() bool {
	// Get the token ahead of the current index.
	possibleIndexEosToken := p.GetCurrentToken().GetTokenIndex() - 1
        if possibleIndexEosToken < 0 {
            return false
        }
	ahead := p.GetTokenStream().Get(possibleIndexEosToken)

	if ahead.GetChannel() != antlr.LexerHidden {
		// We're only interested in tokens on the HIDDEN channel.
		return true
	}

	if ahead.GetTokenType() == JavaScriptParserLineTerminator {
		// There is definitely a line terminator ahead.
		return true
	}

	if ahead.GetTokenType() == JavaScriptParserWhiteSpaces {
		// Get the token ahead of the current whitespaces.
		possibleIndexEosToken = p.GetCurrentToken().GetTokenIndex() - 2
		ahead = p.GetTokenStream().Get(possibleIndexEosToken)
	}

	// Get the token's text and type.
	text := ahead.GetText()
	_type := ahead.GetTokenType()

	// Check if the token is, or contains a line terminator.
	return (_type == JavaScriptParserMultiLineComment && (strings.Contains(text, "\r") || strings.Contains(text, "\n"))) ||
		(_type == JavaScriptParserLineTerminator)
}
